### DISCUSSION TITLE
kube-derive: yaml vs. rust question

### DISCUSSION THREAD

**Discussion by webern** [kube-rs/kube#554] [Q&A] [answered]
Is the correct workflow to maintain a CRD spec both as a YAML file and as a Rust struct with the CustomResource derive? Or can the YAML be generated from the Rust struct (or the Rust struct from the YAML)? Or can the CRD be applied to the cluster by a Rust program that skips the YAML formatting altogether?

Thanks. I'm just getting started. I hope the question is clear. I'm trying to understand the developer workflow for keeping

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: mycrd
```

in sync with

```rust
#[derive(CustomResource, Serialize, Deserialize, Debug, PartialEq, Clone, JsonSchema)]
#[kube(
    group = "example.com",
    version = "v1",
    kind = "MyCrd",
    plural = "mycrds",
    derive = "PartialEq",
    namespaced
)]
#[kube(status = "MyCrdStatus")]
pub struct MyCrdSpec {
  // ...
}
```

**Comment by clux** [comment, marked as answer]
> Is the correct workflow to maintain a CRD spec both as a YAML file and as a Rust struct with the CustomResource derive? Or can the YAML be generated from the Rust struct (or the Rust struct from the YAML)?

You can generate the yaml from the rust struct (not the other way around), and you should commit that generated file, because the definition is usually applied out of band (crd write perms are usually cluster admin type permissions, and you generally don't want your controller to have it).

I recommend a small bin entry point to generate this file like [crdgen in controller-rs](https://github.com/clux/controller-rs/blob/master/src/crdgen.rs).

> Or can the CRD be applied to the cluster by a Rust program that skips the YAML formatting altogether?

You can do this, but for the permission issue, this is generally not good practice.

**Reply by webern** [reply]
Very helpful, thank you!

**Reply by webern** [reply]
I have an additional, related question. I was looking for pattern constraints like this:

```yaml
some_pattern:
  type: string
  pattern: '^\d{3}-\d{2}-\d{4}$'
```

My guess is that this PR https://github.com/GREsau/schemars/pull/78 might make this possible in the future with `kube-derive`?
Am I understanding correctly that `kube-derive` is using schemars to generate a JSON schema then transforming it to produce OpenAPIv3?

**Reply by clux** [reply]
Yeah, that's all correct. There are some ongoing discussions in onwards from https://github.com/clux/kube-rs/issues/129#issuecomment-749463718 about this, but waiting for nice sugar. You can at least get around validation limitations by manually doing implementations of `Schema`. The [crd_derive example is good here](https://github.com/clux/kube-rs/blob/3b46233e7cb019751ff7c6cc085f2914acf00b4c/examples/crd_derive.rs#L34-L78).

**Comment by p-alik** [comment]
[crdgen.rs of controller-rs](https://github.com/kube-rs/controller-rs/blob/main/src/crdgen.rs) generates _YAML_ for the [DocumentSpec](https://github.com/kube-rs/controller-rs/blob/c4aabac479f59c1e05579fd96ee28265ac6c3abe/src/manager.rs#L33-L40).
Is it possible to generate vice versa `struct DocumentSpec` from _YAML_?

**Reply by clux** [reply]
yep! see https://github.com/kube-rs/kopium
